### PR TITLE
sql: Handle empty publications and missing for table clauses in pg sources

### PR DESCRIPTION
**Pull request by moulimukherjee**
Added errors for the following cases

- Trying to use an empty publication when creating a postgres source
- Missing FOR ALL TABLES or FOR TABLES clause when using postgres as a source

### Motivation
Along with PR https://github.com/MaterializeInc/materialize/pull/17923 fixes https://github.com/MaterializeInc/materialize/issues/15499

### Tips for reviewer

Is the wording of the error messages fine?

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql/src/pure.rs [hunk 1]
```diff
@@ -318,6 +318,9 @@ pub async fn purify_create_source(
             let mut validated_requested_subsources = vec![];
             match referenced_subsources {
                 Some(ReferencedSubsources::All) => {
+                    if publication_tables.is_empty() {
+                        sql_bail!("FOR ALL TABLES is only valid for non-empty publication sources");
```
Comment (guswynn) [inline src/sql/src/pure.rs:322]:
```suggestion
                        sql_bail!("FOR ALL TABLES is only valid for non-empty publications");
```

File: src/sql/src/pure.rs [hunk 2]
```diff
@@ -329,6 +332,11 @@ pub async fn purify_create_source(
                     }
                 }
                 Some(ReferencedSubsources::Subset(subsources)) => {
+                    if publication_tables.is_empty() {
+                        sql_bail!(
+                            "FOR TABLES (..) is only valid for non-empty publication sources"
```
Comment (guswynn) [inline src/sql/src/pure.rs:337]: same re-wording as above!

File: src/sql/src/pure.rs [hunk 3]
```diff
@@ -347,7 +355,9 @@ pub async fn purify_create_source(
                     validated_requested_subsources
                         .extend(subsource_gen(subsources, &publication_catalog)?);
                 }
-                None => {}
+                None => {
+                    sql_bail!("FOR ALL TABLES or FOR TABLES (..) must be specified");
```
Comment (guswynn) [inline src/sql/src/pure.rs:359]: copy the wording from below for load counters:
```
sql_bail!("multi-output sources require a FOR TABLES (..) or FOR ALL TABLES statement");
```

**Review by guswynn — COMMENTED** [review]
Looks good! I just have a few nits on wording

Also, can you run all your test statements on `main` without your changes, and then post what `SHOW CREATE SOURCE` does for each in a comment in this pr? That way people coming to pr can easily see that its not a backwards incompatible change!

**Comment by moulimukherjee** [comment]
Pasting the output below

For all tables with empty publication
```
materialize=> create source "mz_source_empty1" from POSTGRES CONNECTION pg (PUBLICATION 'mz_source_empty') for all tables;
CREATE SOURCE
materialize=> show create source mz_source_empty1;
ERROR:  Expected identifier, found right parenthesis
```

For tables with empty publication, it correctly deduced that the mentioned table was not part of the publication. But anyway, this wouldn't have run successfully. With the changes in the PR, the error message would be different this time.
```
materialize=> create source "mz_source_empty2" from POSTGRES CONNECTION pg (PUBLICATION 'mz_source_empty') for tables (t1);
ERROR:  table t1 not found in source
```

Missing FOR TABLES/FOR ALL TABLES clause
```
materialize=> create source "mz_source_empty3" from POSTGRES CONNECTION pg (PUBLICATION 'mz_source_empty');
CREATE SOURCE
materialize=> show create source mz_source_empty3;
ERROR:  Expected identifier, found right parenthesis
```

File: test/pg-cdc/pg-source-errors.td [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
```
Comment (guswynn) [inline test/pg-cdc/pg-source-errors.td:1]: sorry for the final nit, can you rename this file to `empty-publication.td`? thanks!

**Review by guswynn — APPROVED** [review]

File: test/pg-cdc/pg-source-errors.td (hunk 4 above)
Comment (moulimukherjee) [reply test/pg-cdc/pg-source-errors.td:1]: sure!

**Pull request merged**